### PR TITLE
Implement memory usage logging

### PR DESCRIPTION
**Pull request by hainsdominic**
### What are you trying to accomplish?

The run function can now calculate the memory usage of a Function and add it to the FunctionRunResult struct so it can display it. I also added a unit test to check for a stack overflow error using a benchmark Function that allocates 80MB on the stack. The limit of 256KB is imposed by wasmtime.

### What should reviewers focus on?

I checked out @jianghong's branch about memory limit, but I logged it instead of interrupting the Function, because of [this](https://github.com/Shopify/script-runner/issues/10).

#### On Shopify

Easier to test. De-bloats the main.

#### On merchants

No change

#### On third-party apps

No change.

### Tophat 🎩

You can test your own script by running `cargo run --release -- <path/to/input.json> -s <path/to/script.wasm>`.

Note that `--release` is mandatory to benchmark since the runtime is much shorter when the binary is optimized.

Additionally, I wrote automated tests (that are executed in an opt-3 environment), so `cargo test` will show the results of these tests.

## Before you deploy

- [x] I [tophatted](https://development.shopify.io/engineering/developing_at_Shopify/write-code/tophatting) or tested this change.

### CODE REVIEW

**Comment by DuncanUszkay1** [comment]
Should we add the source code for those wasm files?

File: src/engine.rs [hunk 1]
```diff
@@ -43,11 +44,17 @@ pub fn run(script_path: PathBuf, input_path: PathBuf) -> Result<FunctionRunResul
         let start = Instant::now();
 
         let instance = linker.instantiate(&mut store, &module)?;
+
+        let memory = instance
+            .get_memory(&mut store, "memory")
+            .ok_or(anyhow::format_err!("failed to find `memory` export"))?;
```
Comment (DuncanUszkay1) [inline src/engine.rs:50]: Is this a standard? Do we need to document this somewhere?

File: src/engine.rs [hunk 2]
```diff
@@ -100,6 +111,38 @@ mod tests {
...
+    }
+
+    #[test]
+    #[should_panic]
+    fn test_panic() {
+        run(
+            Path::new("tests/benchmarks/stack_overflow.wasm").to_path_buf(),
+            Path::new("tests/benchmarks/stack_overflow.json").to_path_buf(),
+        )
+        .unwrap();
```
Comment (DuncanUszkay1) [inline src/engine.rs:146]: It seems that the run function itself doesn't panic, it returns an `Err` result. We should test for that instead of having a panic in the test that we're looking for.

File: src/engine.rs [hunk 3]
```diff
@@ -81,6 +89,9 @@ mod tests {
     use super::*;
     use std::path::Path;
 
+    // Arbitrary, used to verify that the runner works as expected.
+    const RUNTIME_THRESHOLD: Duration = Duration::from_millis(5);
```
Comment (DuncanUszkay1) [inline src/engine.rs:93]: Let's work that comment into the name. `ARBITRARY_THRESHOLD` should do it. Naturally any test is used to make sure the runner works as expected.

File: src/engine.rs (hunk 1 above)
Comment (hainsdominic) [reply src/engine.rs:50]: Its the way they do it in the docs and in their book: https://docs.rs/wasmtime/0.37.0/wasmtime/struct.Memory.html
https://docs.wasmtime.dev/contributing-architecture.html?highlight=memory#linear-memory

File: src/engine.rs [hunk 4]
```diff
@@ -100,6 +111,38 @@ mod tests {
...
+
+    #[test]
+    fn test_memory_usage_under_threshold() {
+        let function_run_result = run(
+            Path::new("tests/benchmarks/hello_world.wasm").to_path_buf(),
+            Path::new("tests/benchmarks/hello_world.json").to_path_buf(),
+        )
+        .unwrap();
+
+        assert_eq!(function_run_result.memory_usage, 17);
```
Comment (DuncanUszkay1) [inline src/engine.rs:125]: Let's constantize these memory values

File: tests/benchmarks/hello_42_pages.json [hunk 5]
```diff
@@ -0,0 +1,11 @@
+{
+  "input": {
+    "context": {
+      "suffix": "From Core!"
+    }
+  },
+  "configuration": {
+    "message": "From Core!"
+  },
+  "extension_point": "hello_world"
```
Comment (DuncanUszkay1) [inline tests/benchmarks/hello_42_pages.json:10]:
```suggestion
  "function_api": "hello_world"
```
New terminology :)

File: tests/benchmarks/stack_overflow.json [hunk 6]
```diff
@@ -0,0 +1,11 @@
+{
+  "input": {
+    "context": {
+      "suffix": "From Core!"
+    }
+  },
+  "configuration": {
+    "message": "From Core!"
+  },
+  "extension_point": "hello_world"
```
Comment (DuncanUszkay1) [inline tests/benchmarks/stack_overflow.json:10]:
```suggestion
  "function_api": "hello_world"
```

File: src/engine.rs (hunk 2 above)
Comment (DuncanUszkay1) [reply src/engine.rs:146]: Right, it just seems like an extra step that we don't need. We could construct every test by creating an error and unwrapping, but why would we?

File: src/engine.rs (hunk 1 above)
Comment (DuncanUszkay1) [reply src/engine.rs:50]: I don't see any use of the magic `memory` name in those docs 🤔 In fact the first link uses `mem` in some of their examples

Comment (DuncanUszkay1) [reply src/engine.rs:50]: Could we potentially use the [exports function](https://docs.rs/wasmtime/0.37.0/wasmtime/struct.Instance.html#method.exports)?

**Comment by hainsdominic** [comment]
For the test Functions, here is what they do/how I implemented them:
- hello_world, the simple hello world script
- hello_world_42_pages modified the requested number of pages at the end of the WAT file, so it wasn't obtained by compiling code
- sleeps: added a sleep for a duration of 42 seconds in the hello_world script
- stack_overflow: hello_world + allocated a huge array of u64 on the stack (80MB)

I don't think that any of them would be subject to change however

**Comment by jianghong** [comment]
Going to add @KevinRizzoTO for review since he worked on memory usage in our runtime engine. It would be good if this was aligned with how that is measured too

**Comment by KevinRizzoTO** [comment]
> Going to add @KevinRizzoTO for review since he worked on memory usage in our runtime engine. It would be good if this was aligned with how that is measured too

Thanks for the add @jianghong. I actually like this implementation better than the one I ended up with. In short, I used the [ResourceLimiter](https://docs.wasmtime.dev/api/wasmtime/trait.ResourceLimiter.html) trait to hook into calls to the [`memory.grow`](https://developer.mozilla.org/en-US/docs/WebAssembly/Reference/Memory/Grow) instruction. I also used this as a means to get a rough estimation of memory used by saving the `current` value passed to the [`memory_growing`](https://docs.wasmtime.dev/api/wasmtime/trait.ResourceLimiter.html#tymethod.memory_growing) method (see [here](https://github.com/Shopify/runtime-engine/blob/5b57407e10631b9d09da6ee70a7dc952b630ef8d/core/src/shopify_script.rs#L28-L38)). This generally works, but if the script never calls `memory.grow` then we get a reading of zero for memory usage. Not ideal, but my main goal was to get a rough estimate of what our limits should be so we can provide some sane defaults. In that regard, it works well enough for now until I can get the bandwidth to read the memory size directly from the export.

File: src/engine.rs (hunk 1 above)
Comment (KevinRizzoTO) [reply src/engine.rs:50]: I was curious about this as well so took a look. It seems that `memory` is the [standard export name](https://hemingwayapp.com/google.com) when compiling Rust to the `wasm` target. I'm not sure you can change this except if you modified the resulting binary rust gets compiled into. I wouldn't recommend documenting this since it might cause more confusion. Unless the developer is writing a script in pure `WAT` which would be...painful 😆.

A single `wasm` module can [have only one memory](https://webassembly.github.io/spec/core/syntax/modules.html#syntax-mem:~:text=In%20the%20current%20version%20of%20WebAssembly%2C%20at%20most%20one%20memory%20may%20be%20defined%20or%20imported%20in%20a%20single%20module), but many memories is [being proposed](https://github.com/WebAssembly/multi-memory/blob/main/proposals/multi-memory/Overview.md). At that point, I imagine this will need updating.

Comment (DuncanUszkay1) [reply src/engine.rs:50]: >It seems that memory is the [standard export name](https://rustwasm.github.io/docs/book/reference/js-ffi.html#from-the-rust-side:~:text=In%20addition%20to,exported%20as%20%22memory%22) when compiling Rust to the wasm target.

Being a Rust standard is insufficient. ShopifyVM is a Wasm platform, anything we do must work generally for Wasm modules.

**Comment by KevinRizzoTO** [comment]
@DuncanUszkay1 Not at this time. The service is going to be deprecated once we finish the work for inline execution so it's probably not worth it just yet. The tracking of total memory usage isn't user facing, just sent to statsd to graph in datadog

**Comment by DuncanUszkay1** [comment]
>The tracking of total memory usage isn't user facing, just sent to statsd to graph in datadog

Gotcha. As long as we ensure that anything that is given to a developer is reproducible with the script runner I think we're good.

File: src/engine.rs (hunk 1 above)
Comment (DuncanUszkay1) [reply src/engine.rs:50]: Summing up everything they export seems like a reasonable solution,  given that we have access to list of memory export names. I'd rather not document a specific naming limitation if we could remove the limitation instead.

Comment (saulecabrera) [reply src/engine.rs]: This is not a Rust standard, is a standard across compiler infrastructures (LLVM, Binaryen). As Kevin mentioned, you could post-process your Wasm and change the name of the exported memory, not sure what you'd gain by doing so though. Once multiple memories are officially supported by the compiler toolchains I'm pretty sure we'd need to update this. For now this seems a good, standard approach.

Comment (saulecabrera) [reply src/engine.rs]: Also multiple memories is going to introduce some complexities to measure resource usage. I suspect that with multiple memories we'd need to implement the Resource Limiter because not all memories declared in a Wasm module are expected to be exported. So the current approach won't be accurate.

Comment (DuncanUszkay1) [reply src/engine.rs]: >So the current approach won't be accurate.

True, but that's true regardless of whether or not we adopt the `"memory"` name standard. If that's an issue, let's open a separate separate thread about the approach of the PR in general.

In the latest commit Dominic added an approach that sums up all exported memories, which avoids this naming limitation. Is there a reason to re-add the limitation?

Comment (saulecabrera) [reply src/engine.rs]: Yeah, the approach looks good. With the recent changes I don't see the need to reintroduce the "memory" naming standard. I emphasize on multi-memory because when we support it,  for example for module linking, the approach here will need to differ considerably. Just want to make sure that this is on your radar.
